shared_model/builders/protobuf/transaction.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_TRANSACTION_BUILDER_HPP
+#define IROHA_PROTO_TRANSACTION_BUILDER_HPP
+
+#include "backend/protobuf/transaction.hpp"
+
+#include "block.pb.h"
+#include "commands.pb.h"
+
+#include "amount/amount.hpp"
+#include "builders/protobuf/helpers.hpp"
+#include "interfaces/common_objects/types.hpp"
+
+namespace shared_model {
+  namespace proto {
+    template <int S = 0>
+    class TemplateTransactionBuilder {
+     private:
+      template <int>
+      friend class TemplateTransactionBuilder;
+
+      enum RequiredFields { Command, CreatorAccountId, TxCounter, TOTAL };
+
+      template <int s>
+      using NextBuilder = TemplateTransactionBuilder<S | (1 << s)>;
+
+      iroha::protocol::Transaction transaction_;
+
+      template <int Sp>
+      TemplateTransactionBuilder(const TemplateTransactionBuilder<Sp> &o)
+          : transaction_(o.transaction_) {}
+
+     public:
+      TemplateTransactionBuilder() = default;
+
+      NextBuilder<CreatorAccountId> creatorAccountId(
+          const interface::types::AccountIdType &account_id) {
+        transaction_.mutable_payload()->set_creator_account_id(account_id);
+        return *this;
+      }
+
+      NextBuilder<TxCounter> txCounter(Transaction::TxCounterType tx_counter) {
+        transaction_.mutable_payload()->set_tx_counter(tx_counter);
+        return *this;
+      }
+
+      NextBuilder<Command> addAssetQuantity(
+          const interface::types::AccountIdType &account_id,
+          const interface::types::AssetIdType &asset_id,
+          const std::string &amount) {
+        auto command = transaction_.mutable_payload()
```

Comment:
Maybe better to move `add_commands` as a reference field of class?

Reply:
@lebdron [removed it](https://github.com/hyperledger/iroha/commit/39c5efba66c46823badcb0255cf6ef7b442c3afc#diff-91e268aa9408d1ea56d74911608c784bL40) and I thought there's some reason